pkg/rules/helper.go

```diff
@@ -0,0 +1,64 @@
+package rules
+
+import (
+	"bytes"
+	"fmt"
+	"strconv"
+
+	"github.com/thedevsaddam/gojsonq/v2"
+)
+
+// checkSecurityContextFn is the used to check attributes in
+// the securityContext of a pod, initContainer or container. It
+// expects the json node of a pod or a container.
+type checkSecurityContextFn func(*gojsonq.JSONQ) checkSecurityContextResult
+
+// checkSecurityContextResult defines the state of an attribute in security context.
+// This helps making decision with attribute precedence.
+type checkSecurityContextResult struct {
+	// unset defines when an attribute is not set.
+	unset bool
+	// valid defines when an attribute is correctly configured.
+	valid bool
+}
+
+func checkSecurityContext(json []byte, checkPodSecurityContext bool, checkFn checkSecurityContextFn) int {
+	spec := getSpecSelector(json)
+
+	// Some attributes can be set in the PodSecurityContext and
+	// will be common to all containers but only if the same attribute is
+	// not also set in container.securityContext because field values of
+	// container.securityContext take precedence over field values of PodSecurityContext.
+	var attrValidAtPodLevel bool
+	if checkPodSecurityContext {
+		jqPod := gojsonq.New().Reader(bytes.NewReader(json)).From(spec)
+
+		if res := checkFn(jqPod); res.valid {
+			attrValidAtPodLevel = true
+		}
+	}
+
+	containersLen := gojsonq.New().Reader(bytes.NewReader(json)).From(spec + ".containers").Count()
+	initContainersLen := gojsonq.New().Reader(bytes.NewReader(json)).From(spec + ".initContainers").Count()
+
+	var valid int
+
+	ctnFn := func(containerCount int, containerNode string) {
+		for i := 0; i < containerCount; i++ {
+			// e.g: .initContainer.[0]
+			node := fmt.Sprintf(".%s.[%s]", containerNode, strconv.Itoa(i))
+			c := gojsonq.New().Reader(bytes.NewReader(json)).From(spec + node)
```

Comment:
do we have to keep creating new gosjonq stuff all the time?

Reply:
yes because the `containerNode` parameter differs when the function is used.

Reply:
Looks like we can copy the base `gojsonq.New().Reader(bytes.NewReader(json))` and then do all our separate `From`s

https://gojsonq.github.io/#copy

Again this might not actually affect speed or memory footprint but having many separate readers and instances just feels like it would slow down for very large json

Reply:
good point :+1:  I did a quick benchmark before and after the change, this is much faster when creating JQ only once.

before:

```
go test -bench=. -count 5
goos: linux
goarch: amd64
pkg: github.com/controlplaneio/kubesec/v2/pkg/rules
cpu: 12th Gen Intel(R) Core(TM) i7-1260P
BenchmarkCheckSecurityContext-16    	    4770	    237554 ns/op
BenchmarkCheckSecurityContext-16    	    5068	    230770 ns/op
BenchmarkCheckSecurityContext-16    	    5072	    240314 ns/op
BenchmarkCheckSecurityContext-16    	    5107	    238788 ns/op
BenchmarkCheckSecurityContext-16    	    5149	    241402 ns/op
PASS
ok  	github.com/controlplaneio/kubesec/v2/pkg/rules	7.095s
```

after:

```
goos: linux
goarch: amd64
pkg: github.com/controlplaneio/kubesec/v2/pkg/rules
cpu: 12th Gen Intel(R) Core(TM) i7-1260P
BenchmarkCheckSecurityContext-16    	    8612	    130565 ns/op
BenchmarkCheckSecurityContext-16    	    8306	    135635 ns/op
BenchmarkCheckSecurityContext-16    	    8554	    132413 ns/op
BenchmarkCheckSecurityContext-16    	    8655	    135554 ns/op
BenchmarkCheckSecurityContext-16    	    9142	    134617 ns/op
```